[WiP][FIXED JENKINS-33082] Record diagnosis information under a slave disconnection failure

fbelzunc opened the pull request:
This is just a work in progress...

For diagnosis purposes could be nice to record what is happening on the slave side once the pingThread has failed.

@andresrc @oleg-nenashev @jtnord WDYT about this? Might be useful?

jtnord commented on src/main/java/hudson/remoting/Launcher.java:
```diff
@@ -516,19 +516,34 @@ public static void main(InputStream is, OutputStream os, Mode mode, boolean perf
              interval = 1000 * Long.parseLong(
                 System.getProperty("hudson.remoting.Launcher.pingIntervalSec", "600"));
+
+        boolean pingFailureAnalyzer = Boolean.parseBoolean(System.getProperty("hudson.remoting.Launcher.pingFailureAnalyzer", "false"));
```
[`Boolean.getBoolean("hudson.remoting.Launcher.pingFailureAnalyzer")`](https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html#getBoolean(java.lang.String))

jtnord commented on src/main/java/hudson/remoting/Launcher.java:
```diff
@@ -516,19 +516,34 @@ public static void main(InputStream is, OutputStream os, Mode mode, boolean perf
+
         if (performPing && timeout > 0 && interval > 0) {
-            new PingThread(channel, timeout, interval) {
+            new PingThread(channel, timeout, interval, pingFailureAnalyzer) {
```
there are more ways to create/construct the PingThread than just here - why not put this in the base class - or in a class that the PingThread can call?

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+ * Generates diagnosis information (threaddump + stacktrace) when the pingThread fails
+ *
+ * <p>
+ * Useful to troubleshoot the slave disconnection failures on the agent side.
```
should be on both sides - who is to say the fault is not on the masters side?

andresrc replied:
On server-side there is kind of an extension for that. Will have to check,

fbelzunc replied:
@andresrc I think you are talking about http://javadoc.jenkins-ci.org/jenkins/slaves/PingFailureAnalyzer.html and I am already using it on my code. So I guess, this might be useful for master troubleshooting as well.

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+    public void saveStackTrace(Throwable cause) {
+        long iota = System.currentTimeMillis();
+        final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+        File file = stacktraces.file(format.format(new Date(iota++)) + ".txt");
```
what is this iota++?

andresrc replied:
I'd better use the date to identify the case and a different suffix for each diagnosis (stack trace, thread dump, etc), simpler to group the related files.

jtnord commented on src/main/java/hudson/remoting/PingThreadFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public synchronized void add(File f) {
+        if (size <= files.size()) {
```
why the check?

fbelzunc replied:
Because we only want to generate a fix number of troubleshooting files so it doesn't get bigger and bigger. It is just rotating to avoid size issues.

jtnord replied:
so you should be checking if you have too many files then not too few? `if (size >= files.size()) {...`

fbelzunc replied:
No, because you specify a limit with `size`, so if you do `if (size >= files.size())` then you will be deleting from the beginning - so any file will be actually created.

jtnord replied:
must have been late when I commented :-(

jtnord commented on src/main/java/hudson/remoting/PingThread.java:
```diff
@@ -61,14 +61,24 @@
+     */
+    private final boolean pingFailureAnalyzer;
+
+    public PingThread(Channel channel, long timeout, long interval, boolean pingFailureAnalyzer) {
```
don't we want this on by default?

andresrc replied:
Depending on whether we want to change current default behavior.

fbelzunc replied:
I would prefer to have it disabled in order ensure any bad effect of this PR on the remoting behaviour. In any case, I am happy to do whatever you think might be the best option here.

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+        try {
+            long cpuTime = mbean.getThreadCpuTime(t.getThreadId());
+            long threadUserTime = mbean.getThreadUserTime(t.getThreadId());
+            cpuPercentage = (cpuTime == 0) ? 0: 100 * threadUserTime / cpuTime;
```
this is a bug copied from the support core plugin - please don't duplicate it :-)

100% cpu time is for things that have been running without stopping since Jenkins started.
a thread that wakes up every day and runs for 1ms should not be listed at 100%

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+        }
+
+        writer.println();
+        writer.flush();
```
are you are going to close this writer somewhere ;)

andresrc replied:
+1

jtnord replied:
:bug: writer is not closed.

jtnord commented on src/main/java/hudson/remoting/Launcher.java:
```diff
@@ -516,19 +516,34 @@ public static void main(InputStream is, OutputStream os, Mode mode, boolean perf
+
+                @Override
+                protected void onDead(Throwable diagnosis) {
+                    if (this.isPingFailureAnalyzer()) {
```
Will this will spam the logs for orderly shutdown?

https://github.com/jenkinsci/remoting/pull/74/files#diff-ba801c8ba3eabb55537cfff979bb3655L117

fbelzunc replied:
I don't understand this concern.

* Can you elaborate a bit more, please?

jtnord replied:
if you shutdown the launcher (with a ctrl-c, service stop etc) - I assume the channel will be closed and `onDead` will be called back.
If this is true then this seems to causing excessive logging.

andresrc commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+            e.printStackTrace();
+        }
+        cause.printStackTrace(writer);
+        writer.close();
```
`close` in a `finally`

andresrc replied:
And possible NPE.

andresrc commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+
```
`writer` can be `null` past this point.

andresrc commented:
Besides line comments I would like to check the master-side use of the PingThread.

oleg-nenashev commented:
@fbelzunc Any plans to finish it?

fbelzunc commented:
@oleg-nenashev Tomorrow, I promise to go through all the problems pointed by everybody.

fbelzunc commented:
Taking a look at this now.

fbelzunc commented:
@andresrc @jtnord @oleg-nenashev What do you think now? I think I simplified the code.

I would like to fully test the feature, but it would be nice if I can have your insights about this proposition.

Many thanks!

andresrc commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+            cause.printStackTrace(writer);
+        }
+    }
+
```
:bug: method is wrong, the content is written after closing the writer.

andresrc commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+        writer.flush();
+        writer.close();
+    }
+
```
Bad exception handling. In general:

```java
Resource res = getResource();
try {
    // use resource
} catch(InterestingException e) {
    // if any
} finally {
    // free resource
}

andresrc commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+
+        if (!folder.exists() && !folder.mkdirs())
+            throw new Error("Failed to create "+folder);
+
```
What if `file` exists but is not a directory?

fbelzunc replied:
Then, `folder.mkdirs()` will fail

https://docs.oracle.com/javase/7/docs/api/java/io/File.html#mkdirs()

andresrc commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+                return 0;
+            }
+        });
+
```
If the files are created by the system, with an specific pattern, maybe you can rely on it instead of `lastModified` which may provide surprises in some contexts.

andresrc commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+        files.remove(f);
+        add(f);
+    }
+
```
where is this used?

andresrc commented:
In general I would go for somewhat different design. The `AgentFailuresFileListCap` would provide the `Writer` at incident time and when disposing it would simply maintain the desired number of files.

batmat commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+    /**
+     * The cause of the ping failure
+     */
+    private Throwable cause;
```
Possibly make those fields final if the goal is to make that analyzing class immutable

batmat commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+            throw e;
+        } catch (UnsupportedEncodingException e) {
+            throw e;
+        }
```
I must be missing something. Is that whole try/catch block even useful?

batmat commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+        final MonitorInfo[] monitors = t.getLockedMonitors();
+
+        for (int i = 0; i < elements.length; i++) {
+            final StackTraceElement element = elements[i];
```
NIT(?): why not java5 style? Probably leaner.
```
for(StackTraceElement element : elements) {
``` 
instead of those two lines and the `i` index "management".

batmat commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+            final StackTraceElement element = elements[i];
+            writer.printf("    at %s\n", element);
+            for (int j = 1; j < monitors.length; j++) {
+                final MonitorInfo monitor = monitors[j];
```
Same here, not need to have to cope with `i` and `j`.

batmat commented on src/main/java/hudson/remoting/AgentFailuresAnalyzer.java:
```diff
@@ -0,0 +1,220 @@
+            writer.printf("    at %s\n", element);
+            for (int j = 1; j < monitors.length; j++) {
+                final MonitorInfo monitor = monitors[j];
+                if (monitor.getLockedStackDepth() == i) {
```
Ok, you're using `i` there at least.

batmat commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+        this.size = size;
+
+        if (!folder.exists() && !folder.mkdirs())
+            throw new Error("Failed to create "+folder);
```
NIT: just me, but IMO never ever use blocks without braces. Just not worth the risk when touching that code later.

batmat commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+                long l = o1.lastModified() - o2.lastModified();
+                if (l<0)    return -1;
+                if (l>0)    return 1;
+                return 0;
```
replace the 4 lines above by `return o1.lastModified() - o2.lastModified()` ?

batmat commented on src/main/java/hudson/remoting/AgentFailuresFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    /**
+     * Creates a new file object in this directory without changing the
```
"the ..." missing word?

jtnord commented on src/main/java/hudson/remoting/PingThread.java:
```diff
@@ -61,14 +61,24 @@
     }
 
+    public PingThread(Channel channel, long timeout, long interval) {
+        this(channel, 4*60*1000/*4 mins*/, interval, false);
```
:bug: this should pass the timeout supplied and not replace it with something else...

jtnord commented on src/main/java/hudson/remoting/PingThreadFileListCap.java:
```diff
@@ -0,0 +1,83 @@
+ * Useful to record incidents as files without bankrupting
+ *
+ */
+public class PingThreadFileListCap {
```
does this need to be public - seems its internal and there is IIRC something in Jenkins to perform this for other users.

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+    public void takeThreadDump() {
+        long iota = System.currentTimeMillis();
+        final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+        File file = threadDumps.file(format.format(new Date(iota++)) + ".txt");
```
I do not see a point in increementing iota - it is not used after here, and it is not used iin a loop - so on next call you will just get something new in time - unless this happens within a 10ms granualatity on a low resolution timer - but if that is the case the iota++ will still not do anything!

jtnord commented on src/main/java/hudson/remoting/PingFailureAnalyzer.java:
```diff
@@ -0,0 +1,225 @@
+     * Dumps all of the threads' current information to an output stream.
+     *
+     */
+    @IgnoreJRERequirement
```
this seems to be wrong - why are you ignoring animal sniffer results - and what was the cause of the check failing?

christ66 commented:
FWIW I am :ant: on this issue.

* There is already an extension point for analyzing ping failures called `PingFailureAnalyzer`. I tried creating https://github.com/jenkinsci/jenkins/pull/1988 earlier which performed a thread dump on the node machine if it hit a ping timeout.
* I believe this should not be a JVM argument but some checkbox which can be enabled/disabled this option via the UI for all slaves.
* I'm more inclined to have this be a plugin which uses the `PingFailureAnalyzer` extension point. Sometimes thread dump information is not needed, and it might be good to implement more diagnostic tools other than thread dumps (like heap histogram, or networking checks).

fbelzunc commented:
@christ66 Answering you in-line:

> There is already an extension point for analyzing ping failures called PingFailureAnalyzer. 

I am aware of `PingFailureAnalyzer` - I am using it on the PR. But this is not enough to correctly troubleshooting the slave disconnection issues. 

> I tried creating jenkinsci/jenkins#1988 earlier which performed a thread dump on the node machine if it hit a ping timeout.

I am happy to close this PR if you plan to address this problem in short/med-term.

> I believe this should not be a JVM argument but some checkbox which can be enabled/disabled this option via the UI for all slaves.

If this is a concern, it can be addressed.

> I'm more inclined to have this be a plugin which uses the PingFailureAnalyzer extension point. Sometimes thread dump information is not needed, and it might be good to implement more diagnostic tools other than thread dumps (like heap histogram, or networking checks).

I agree my initial plan was to use this only for slaves - as we cannot address the problem with plugins. For Jenkins Master I agree it should be done through support-core plugin.

@oleg-nenashev @christ66 @batmat I have very limited time to work on these matters so before continuing here I would like to ensure that this PR will be fine addressing the comments. So basically create a checkbox on the UI and make this only available for slaves - for master will be support-core plugin.

christ66 commented on src/main/java/hudson/remoting/Launcher.java:
```diff
@@ -516,19 +518,40 @@ public static void main(InputStream is, OutputStream os, Mode mode, boolean perf
+
         if (performPing && timeout > 0 && interval > 0) {
             new PingThread(channel, timeout, interval) {
+                @Deprecated
```
It would be nice to have javadoc for deprecating a method. Something like:

```java
@Deprecated Use {@link PingThread#onDead(Throwable)}
```

oleg-nenashev commented:
I don't like the current design much as well. Makes sense to take it offline and review together

oleg-nenashev commented:
@fbelzunc Any plans to work on it?

fbelzunc commented:
@oleg-nenashev since you don't like the current design, I am transitioning this to close.
